### PR TITLE
Add `entry_or_insert_with_default` lint

### LINKED ISSUE #3812: Suggest `Entry::or_default` for `Entry::or_insert_with(Default::default)`
For code that calls `.or_insert_with(Default::default)` on a `std::collections::btree_map::Entry` or `std::collections::hash_map::Entry`, clippy should suggest [`std::collections::btree_map::Entry::or_default`](https://doc.rust-lang.org/std/collections/btree_map/enum.Entry.html#method.or_default) and [`std::collections::hash_map::Entry::or_default`](https://doc.rust-lang.org/std/collections/hash_map/enum.Entry.html#method.or_default)

### PR DESCRIPTION
**Pull request by cauebs**
Closes #3812 

Checks for calls to `.or_insert_with` on `Entry` (for `BTreeMap` or `HashMap`) with `Default::default` or equivalent as the argument, and suggests `.or_default`.

Feedback is welcome!

changelog: none

### CODE REVIEW

**Comment by cauebs** [comment]
Had some trouble with the documentation test. Still not sure how to run them locally... All good now.

File: clippy_lints/src/methods/mod.rs [hunk 1]
```diff
@@ -1709,6 +1753,28 @@ impl<'tcx> LateLintPass<'tcx> for Methods {
...
+                span,
+                "use of `or_insert_with` with `Default::default`",
+                "replace with",
+                "or_default()".to_owned(),
+                Applicability::MachineApplicable,
+            );
+        }
+    }
+}
+
```
Comment (camsteffen) [inline clippy_lints/src/methods/mod.rs:1777]: I wonder if this function should be merged with `check_unwrap_or_default` a few lines below

Comment (cauebs) [reply clippy_lints/src/methods/mod.rs:1777]: Not sure, but I would need some help figuring out how, to be honest.

**Comment by cauebs** [comment]
@Manishearth can you just add the `hacktoberfest-accepted` label to this PR before monday? We can iron things out later if necessary.

File: clippy_lints/src/methods/mod.rs [hunk 2]
```diff
@@ -1515,6 +1556,9 @@ impl<'tcx> LateLintPass<'tcx> for Methods {
             ["unwrap_or_else", ..] => unnecessary_lazy_eval::lint(cx, expr, arg_lists[0], "unwrap_or"),
             ["get_or_insert_with", ..] => unnecessary_lazy_eval::lint(cx, expr, arg_lists[0], "get_or_insert"),
             ["ok_or_else", ..] => unnecessary_lazy_eval::lint(cx, expr, arg_lists[0], "ok_or"),
+            ["or_insert_with", ..] => {
+                lint_entry_or_insert_with_default(cx, method_spans[0].with_hi(expr.span.hi()), arg_lists[0])
```
Comment (Manishearth) [inline clippy_lints/src/methods/mod.rs:1560]: I don't think you should need to do span splicing

File: clippy_lints/src/methods/mod.rs (hunk 1 above)
Comment (Manishearth) [reply clippy_lints/src/methods/mod.rs:1777]: Yes, this should be merged. You'll want to hoist that function out and make the `unwrap_or` part a parameter

**Review by Manishearth — APPROVED** [review]
We don't use that label, but marking this as approved should be enough

File: clippy_lints/src/methods/mod.rs (hunk 2 above)
Comment (cauebs) [reply clippy_lints/src/methods/mod.rs:1560]: Can you elaborate? I confess I didn't quite understand this part and just tried to adapt what was being done for the others.

Comment (Manishearth) [reply clippy_lints/src/methods/mod.rs:1560]: Actually, never mind, some of the others do span splicing too. I just don't think the splicing should be _here_, but if you're merging this with the `check_unwrap_or_default` thing then it's going to go away from here anyway.

**Comment by giraffate** [comment]
ping from triage @cauebs. There seems to be a small fix left to be done. Do you have any questions on how to proceed here?

**Comment by cauebs** [comment]
I believe it needs a small rewrite, and I haven't had the time to do it. Next week I'll probably be able to get this done. Sorry for the delay.

**Comment by giraffate** [comment]
ping from triage @cauebs. Sorry for pinging again, but do you have any update on this?

**Comment by giraffate** [comment]
ping from triage @cauebs. Thanks for contributing Clippy! Sadly this PR didn't have any update in the last 2 weeks. If you have more time to continue working on this, feel free to reopen.

**Pull request closed without merging**